peeklock/renewer.go

```diff
@@ -0,0 +1,54 @@
+package peeklock
+
+import (
+	"context"
+	"time"
+
+	servicebus "github.com/Azure/azure-service-bus-go"
+)
+
+type LockRenewer interface {
+	RenewLocks(ctx context.Context, messages ...*servicebus.Message) error
+}
+
+type PeriodicLockRenewer struct {
+	lockRenewer LockRenewer
+	cancelFunc  func()
+	timer       *time.Timer
+}
+
+func RenewPeriodically(ctx context.Context, interval time.Duration, lockrenewer LockRenewer, msg *servicebus.Message) *PeriodicLockRenewer {
+	p := &PeriodicLockRenewer{
+		lockRenewer: lockrenewer,
+	}
+	go p.startPeriodicRenewal(ctx, interval, msg)
+	return p
+}
+
+func (plr *PeriodicLockRenewer) startPeriodicRenewal(ctx context.Context, interval time.Duration, message *servicebus.Message) {
+	var renewerCtx context.Context
+	renewerCtx, plr.cancelFunc = context.WithCancel(ctx)
+	for alive := true; alive; {
+		select {
+		case <-time.After(interval):
+			// add span for logs
+			// fmt.Printf("%s - Renewing Lock: [ ttl: %s, lockedUntil: %s, enqueuedTime: %s ]",
+			// 	time.Now(),
+			// 	message.TTL,
+			// 	message.SystemProperties.LockedUntil,
+			// 	message.SystemProperties.EnqueuedTime)
+			plr.lockRenewer.RenewLocks(renewerCtx, message)
+			// what do we do if we fail to renew?
```

Comment:
will it be too aggressive to discard the msg if renew lock failed? the msg might be picked up by another thread if we still keep it after lock renew failed

another question might be more related to ttl, I am curious if we trying to complete a msg after ttl, will it simply failed or even blocked the thread?

Reply:
I think that's a good idea. If we fail to renew the lock, just fail
The message.
The underlying rpc call has retries already, so if we fail here, something's wrong. Thanks 👍

Reply:
Ah renewlocks already has retries? Do you know how much it backs off? My original thought was to keep retrying till we got to lock duration but internal retries makes that dubious

My prefernce if we fail after retries is to cancel the cotnext we gave to the user provideded handler so they can stop if need be.

Reply:
under the hood, renewLocks does : 
```
client, err := re.entity.GetRPCClient(ctx)
	if err != nil {
		tab.For(ctx).Error(err)
		return err
	}
	return client.RenewLocks(ctx, messages...)
```

the rpcclient always does (minus error handling for simplicity) : 

```
if err := r.ensureConn(ctx); err != nil {
		tab.For(ctx).Error(err)
		return err
	}
...
rpcLink, err := rpc.NewLink(r.client, r.ec.ManagementPath())
... 
response, err := rpcLink.RetryableRPC(ctx, 3, 1*time.Second, renewRequestMsg)
...
```

Reply:
with retryableRPC signature being: 

```
func (l *Link) RetryableRPC(ctx context.Context, times int, delay time.Duration, msg *amqp.Message) (*Response, error) {
```